src/components/WideRHPContextProvider/index.tsx

```diff
@@ -91,77 +127,189 @@ const expenseReportSelector = (reports: OnyxCollection<Report>) => {
+    const [superWideRHPRouteKeys, setSuperWideRHPRouteKeys] = useState<string[]>([]);
     const [shouldRenderSecondaryOverlay, setShouldRenderSecondaryOverlay] = useState(false);
+    const [shouldRenderTertiaryOverlay, setShouldRenderTertiaryOverlay] = useState(false);
     const [expenseReportIDs, setExpenseReportIDs] = useState<Set<string>>(new Set());
+    const [multiTransactionExpenseReportIDs, setMultiTransactionExpenseReportIDs] = useState<Set<string>>(new Set());
+    const [isWideRHPClosing, setIsWideRHPClosing] = useState(false);
+    const focusedRouteKey = useRootNavigationState((state) => (state ? findFocusedRoute(state)?.key : undefined));
     const [allReports] = useOnyx(ONYXKEYS.COLLECTION.REPORT, {selector: expenseReportSelector, canBeMissing: true});
 
-    // Return undefined if RHP is not the last route
-    const lastVisibleRHPRouteKey = useRootNavigationState((state) => {
-        // Safe handling when navigation is not yet initialized
-        if (!state) {
-            return undefined;
-        }
-        const lastFullScreenRouteIndex = state?.routes.findLastIndex((route) => isFullScreenName(route.name));
-        const lastRHPRouteIndex = state?.routes.findLastIndex((route) => route.name === NAVIGATORS.RIGHT_MODAL_NAVIGATOR);
+    const syncWideRHPKeys = useCallback(() => {
+        setWideRHPRouteKeys(getCurrentWideRHPKeys(allWideRHPRouteKeys, getLastVisibleRHPRouteKey(navigationRef.getRootState())));
+    }, [allWideRHPRouteKeys]);
 
-        // Both routes have to be present and the RHP have to be after last full screen for it to be visible.
-        if (lastFullScreenRouteIndex === -1 || lastRHPRouteIndex === -1 || lastFullScreenRouteIndex > lastRHPRouteIndex) {
-            return undefined;
-        }
+    const syncSuperWideRHPKeys = useCallback(() => {
+        setSuperWideRHPRouteKeys(getCurrentWideRHPKeys(allSuperWideRHPRouteKeys, getLastVisibleRHPRouteKey(navigationRef.getRootState())));
+    }, [allSuperWideRHPRouteKeys]);
 
-        return state?.routes.at(lastRHPRouteIndex)?.key;
-    });
+    const clearWideRHPKeys = useCallback(() => {
+        setWideRHPRouteKeys([]);
+        setSuperWideRHPRouteKeys([]);
+    }, []);
 
-    const wideRHPRouteKeys = useMemo(() => {
-        const rootState = navigationRef.getRootState();
+    useEffect(() => {
+        syncWideRHPKeys();
+    }, [allWideRHPRouteKeys, syncWideRHPKeys]);
 
-        if (!rootState) {
-            return [];
-        }
+    useEffect(() => {
+        syncSuperWideRHPKeys();
+    }, [allSuperWideRHPRouteKeys, syncSuperWideRHPKeys]);
 
-        const lastRHPRoute = rootState.routes.find((route) => route.key === lastVisibleRHPRouteKey);
+    const isWideRHPFocused = useMemo(() => {
+        return !!focusedRouteKey && wideRHPRouteKeys.includes(focusedRouteKey);
+    }, [focusedRouteKey, wideRHPRouteKeys]);
 
-        if (!lastRHPRoute) {
-            return [];
-        }
+    const getIsSuperWideRHPBelowFocusedScreen = useCallback(
```

Comment:
can you please add a comment here explaining what the value represents/ what it is for in plain english